### ISSUE TITLE
Publish coursier for Scala 3

### ISSUE THREAD

**Issue by thesamet** [coursier/coursier#2001]
ScalaPB uses Coursier as a library. It is cross-compiled for Scala 3 using `withDottyCompat`. Since Coursier depends on `scala-xml` , we are running into an issue updating scalatest to the latest version. See https://github.com/scalatest/scalatest/issues/1973

**Comment by alexarchambault** [comment]
@thesamet You should be able to get rid of that error with exclusions. Excluding `org.scala-lang.modules:scala-xml_2.13` from the coursier dependency should work.

About publishing coursier for Scala 3, this is going to take some time, as coursier heavily relies on [data-class](https://github.com/alexarchambault/data-class). data-class uses macro annotations, which aren't available in Scala 3. Code bases using [simulacrum](https://github.com/typelevel/simulacrum) had a similar problem.

My plan is to implement https://github.com/alexarchambault/data-class/issues/27, then use that to generate Scala 3 specific sources in coursier. Implementing https://github.com/alexarchambault/data-class/issues/27 is probably going to take some time, so you'll have to keep using the coursier 2.13 artifacts for a while.

**Comment by kubukoz** [comment]
I'm doing the exclusion thing and apparently there's also a dependency on `com.github.plokhotnyuk.jsoniter-scala:jsoniter-scala-core`. It used to work fine, but I've been [getting linkage errors since 2.1.15](https://github.com/kubukoz/smithy-playground/pull/487): `java.lang.NoClassDefFoundError: com/github/plokhotnyuk/jsoniter_scala/core/JsonValueCodec$mcZ$sp`

Do you know of a workaround? Asking here because this is all due to the `CrossVersion.for3Use2_13` trick.

**Comment by adpi2** [comment]
It seems the problem is `jsoniter-scala-core` uses `@specialized` which is not supported by Scala 3. So the Scala 2 and Scala 3 versions of `jsoniter-scala-core` are binary incompatible: the specialized versions of `JsonValueCodec` are missing from the Scala 3 jar.

Have you tried setting `CrossVersion.for3Use2_13` on `jsoniter-scala-core`?

**Comment by kubukoz** [comment]
That seems to work for now, yeah: https://github.com/kubukoz/smithy-playground/pull/505

although I'm spooked a bit by the fact that the macros module depends on `core` 😬 it works for now (tests exercise that code path), but are 3.x libraries supposed to be able to read 2.13 signatures without TASTY?

**Comment by SethTisue** [comment]
> are 3.x libraries supposed to be able to read 2.13 signatures without TASTY?

it's other direction that requires TASTy

**Comment by adpi2** [comment]
> but are 3.x libraries supposed to be able to read 2.13 signatures without TASTY?

Yes and it is enabled by default. That's what we always do to load the Scala 2.13 library in Scala 3. It has proven to be more stable than the TASTy reader in Scala 2.13 which is disabled by default.

**Comment by kubukoz** [comment]
Just curious, would it be feasible to replace data-class with something like [contraband](https://github.com/sbt/contraband), instead of using scalafix to rewrite the dataclasses once? I'm assuming the scalafix thing would be one-time and further changes to the datatypes would require manual maintenance of the APIs.
